Src/Newtonsoft.Json/JsonTextReader.Async.cs

```diff
@@ -0,0 +1,2327 @@
+        {
+            return SafeAsync ? DoReadAsync(cancellationToken) : base.ReadAsync(cancellationToken);
+        }
+
+        internal async Task<bool> DoReadAsync(CancellationToken cancellationToken)
+        {
+            EnsureBuffer();
+
+            for (;;)
+            {
+                switch (_currentState)
+                {
+                    case State.Start:
+                    case State.Property:
+                    case State.Array:
+                    case State.ArrayStart:
+                    case State.Constructor:
+                    case State.ConstructorStart:
+                        return await ParseValueAsync(cancellationToken).ConfigureAwait(false);
+                    case State.Object:
+                    case State.ObjectStart:
+                        return await ParseObjectAsync(cancellationToken).ConfigureAwait(false);
+                    case State.PostValue:
+
+                        // returns true if it hits
+                        // end of object or array
+                        if (await ParsePostValueAsync(cancellationToken).ConfigureAwait(false))
+                        {
+                            return true;
+                        }
+
+                        break;
+                    case State.Finished:
+                        if (await EnsureCharsAsync(0, false, cancellationToken).ConfigureAwait(false))
+                        {
+                            await EatWhitespaceAsync(false, cancellationToken).ConfigureAwait(false);
+                            if (_isEndOfFile)
+                            {
+                                SetToken(JsonToken.None);
+                                return false;
+                            }
+
+                            if (_chars[_charPos] == '/')
+                            {
+                                await ParseCommentAsync(true, cancellationToken).ConfigureAwait(false);
+                                return true;
+                            }
+
+                            throw JsonReaderException.Create(this, "Additional text encountered after finished reading JSON content: {0}.".FormatWith(CultureInfo.InvariantCulture, _chars[_charPos]));
+                        }
+
+                        SetToken(JsonToken.None);
+                        return false;
+                    default:
+                        throw JsonReaderException.Create(this, "Unexpected state: {0}.".FormatWith(CultureInfo.InvariantCulture, CurrentState));
+                }
+            }
+        }
+
+        private Task<int> ReadDataAsync(bool append, CancellationToken cancellationToken)
+        {
+            return ReadDataAsync(append, 0, cancellationToken);
+        }
+
+        private async Task<int> ReadDataAsync(bool append, int charsRequired, CancellationToken cancellationToken)
+        {
+            if (_isEndOfFile)
```

Comment:
I shared from slightly after this point, as leaving this check-and-return duplicated kept the shared method simpler, and this remaining bit isn't a very large duplication.